### PR TITLE
BACK-965 - Refacto Flag Inputs and Outputs

### PR DESCRIPTION
**Pull request by estrauser-ledger**
## What is this about?

Big refacto : now operations and transactions are separate.
Also flag inputs and outputs belonging to account.

## Cute picture of lama
doublellama

### CODE REVIEW

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Interpreter.scala [hunk 1]
```diff
@@ -1,115 +1,52 @@
...
 
-  def saveTransactions(request: SaveTransactionsRequest, ctx: Metadata): IO[Empty] = {
-
-    //TODO reorg : remove all blocks with height >= minimum transactions block height
-
-    request.transactions.toList
-      .traverse(requestTx => saveTx(request.accountId, requestTx))
-      .map(_ => Empty())
-  }
+  val tranctionInterpreter = new TransactionInterpreter(db)
```
Comment (jjayet-ledger) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Interpreter.scala:19]: `
val transactionInterpreter = new TransactionInterpreter(db)
`

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/OperationInterpreter.scala [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    (for {
+      // Flag inputs with known addresses
+      inputs <- OperationQueries.flagInputs(accountId, addresses.map(_.accountAddress))
+      // Flag outputs with known addresses and update address type (INTERNAL / EXTERNAL)
+      outputs <-
+        addresses
+          .traverse { address =>
+            OperationQueries.flagOutputsForAddress(accountId, address)
+          }
+    } yield (inputs :: outputs).sum).transact(db)
```
Comment (jjayet-ledger) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/OperationInterpreter.scala:88]: I do find this style a little ugly.
I'd prefer a variable but it's more of an aesthetic choice :)

**Review by jjayet-ledger — CHANGES_REQUESTED** [review]

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Interpreter.scala (hunk 1 above)
Comment (estrauser-ledger) [reply coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Interpreter.scala:19]: lol :+1:

File: coins/bitcoin/interpreter/src/main/resources/db/migration/V1__bitcoin_schema.sql [hunk 3]
```diff
@@ -1,8 +1,13 @@
...
 );
 
+CREATE TYPE change_type as ENUM(
+    'internal',
+    'external'
+);
+
 CREATE TABLE block (
     hash VARCHAR NOT NULL PRIMARY KEY,
     height INTEGER NOT NULL,
```
Comment (jjayet-ledger) [inline coins/bitcoin/interpreter/src/main/resources/db/migration/V1__bitcoin_schema.sql:13]: This is declared as BitInt (long) elsewhere.
Either we change this type to BitInt or every other types to Int

**Review by jjayet-ledger — CHANGES_REQUESTED** [review]

File: coins/bitcoin/common/src/main/scala/co/ledger/lama/bitcoin/common/models/Explorer.scala [hunk 4]
```diff
@@ -1,26 +1,20 @@
...
   implicit val config: Configuration = Configuration.default.withSnakeCaseMemberNames
 
-  type BlockHash   = String
-  type BlockHeight = Long
-
   @ConfiguredJsonCodec case class Block(
-      hash: BlockHash,
-      height: BlockHeight,
+      hash: String,
+      height: Long,
```
Comment (estrauser-ledger) [inline coins/bitcoin/common/src/main/scala/co/ledger/lama/bitcoin/common/models/Explorer.scala:15]: BigInt

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/TransactionQueries.scala [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+object TransactionQueries {
+
+  implicit val bigIntType: Meta[BigInt] = Meta.Advanced.other[BigInt]("bigint")
+
+  implicit val operationTypeMeta: Meta[OperationType] =
+    pgEnumStringOpt("operation_type", OperationType.fromKey, _.toString.toLowerCase())
+
+  implicit val changeTypeMeta: Meta[ChangeType] =
+    pgEnumStringOpt("change_type", ChangeType.fromKey, _.toString.toLowerCase())
+
```
Comment (estrauser-ledger) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/TransactionQueries.scala:22]: no need

**Review by nukky94 — APPROVED** [review]

**Review by jjayet-ledger — APPROVED** [review]

**Pull request merged**